### PR TITLE
chrore: parse cgroup v2

### LINKED ISSUE #3812: Failed in deducing any cgroup limits when the instance starts
**Describe the bug**
When the instance starts, a deduce cgroup limit error is reported.

**Expected behavior**

**Screenshots**
image

**Environment (please complete the following information):**
 - OS: [Debian GNU/Linux 11]
 - Kernel: Linux  5.10.0-30-amd64 #1 SMP Debian 5.10.218-1 (2024-06-01) x86_64 GNU/Linux
 - Containerized?: [Bare Metal, Docker, Docker Compose, Docker Swarm, Kubernetes, Other]
 - Dragonfly Version: [v1.23.0]

**Additional context**
When the instance starts, a deduce cgroup limit error is reported.  The directory that reports the error is consistent and the directory exists. At first I thought I didn't have enough permissions, but then I tried to start it with sudo permissions, and the result was still the same. The error doesn't affect the startup of the instance, and the subsequent use of the instance is normal, but I'm not sure if it does or not, or if it only affects it in certain scenarios. If it doesn't, may I suggest changing this to warning or something else? Of course, if there is a problem with the deduce logic, it would be perfect to fix the error output. 😄

### PR DESCRIPTION
**Pull request by kostasrim**
Currently, parsing `cgroups` only works for version 1. In this version, each controller has its own subfolder so for example when we want to fetch memory controller info we search for files in `/sys/fs/cgroup/memory`. However, in `cgroup version 2`, all of the controllers reside in the same folder `/sys/fs/cgroup/`. This PR add support to parse `version 2 cgroups` under the unified directory model. Furthermore, we also not parse `sys/fs/cgroup/user.slice/`

* add support for parsing cgroup version 2
* add `sys/fs/cgroup/user.slice/`

Probably resolves #3812

### CODE REVIEW

File: src/server/dfly_main.cc [hunk 1]
```diff
@@ -462,15 +462,24 @@ bool UpdateResourceLimitsIfInsideContainer(io::MemInfoData* mdata, size_t* max_t
...
-  constexpr auto base_mem = "/sys/fs/cgroup/memory"sv;
-  read_mem(StrCat(base_mem, "/memory.limit_in_bytes"), &mdata->mem_total);
-  read_mem(StrCat(base_mem, "/memory.max"), &mdata->mem_total);
+  auto parse_limits = [&](std::string_view base_mem) {
+    read_mem(StrCat(base_mem, "/memory.limit_in_bytes"), &mdata->mem_total);
+    read_mem(StrCat(base_mem, "/memory.max"), &mdata->mem_total);
+  };
+
+  // For v1
+  constexpr auto base_mem_v1 = "/sys/fs/cgroup/memory"sv;
```
Comment (kostasrim) [inline src/server/dfly_main.cc:471]: This is wrong. There are no `memory.max` in `root cgroup`. That means that always, even before this PR the `read_mem(StrCat(base_mem, "/memory.limit_in_bytes"), &mdata->mem_total);` would never read anything

**Comment by romange** [comment]
How can I test the changes? How to reproduce running under cgroups v2 on my dev machine?

**Comment by romange** [comment]
that's what I get when running from master branch with vlog=1:
```
I20241007 14:32:51.190955 825726 dfly_main.cc:459] mem_path = /sys/fs/cgroup//user.slice/user-1000.slice/session-c1.scope
I20241007 14:32:51.190960 825726 dfly_main.cc:460] cpu_path = /sys/fs/cgroup//user.slice/user-1000.slice/session-c1.scope
I20241007 14:32:51.190971 825726 dfly_main.cc:438] container limits: read /sys/fs/cgroup/memory/memory.limit_in_bytes: N/A
I20241007 14:32:51.190976 825726 dfly_main.cc:438] container limits: read /sys/fs/cgroup/memory/memory.max: N/A
I20241007 14:32:51.190984 825726 dfly_main.cc:438] container limits: read /sys/fs/cgroup//user.slice/user-1000.slice/session-c1.scope/memory.limit_in_bytes: N/A
I20241007 14:32:51.190994 825726 dfly_main.cc:438] container limits: read /sys/fs/cgroup//user.slice/user-1000.slice/session-c1.scope/memory.max: max
I20241007 14:32:51.191002 825726 dfly_main.cc:438] container limits: read /sys/fs/cgroup//user.slice/user-1000.slice/session-c1.scope/memory.high: max

```

**Comment by kostasrim** [comment]
@romange 

> that's what I get when running from master branch with vlog=1:
> 
> ```
> I20241007 14:32:51.190955 825726 dfly_main.cc:459] mem_path = /sys/fs/cgroup//user.slice/user-1000.slice/session-c1.scope
> I20241007 14:32:51.190960 825726 dfly_main.cc:460] cpu_path = /sys/fs/cgroup//user.slice/user-1000.slice/session-c1.scope
> I20241007 14:32:51.190971 825726 dfly_main.cc:438] container limits: read /sys/fs/cgroup/memory/memory.limit_in_bytes: N/A
> I20241007 14:32:51.190976 825726 dfly_main.cc:438] container limits: read /sys/fs/cgroup/memory/memory.max: N/A
> I20241007 14:32:51.190984 825726 dfly_main.cc:438] container limits: read /sys/fs/cgroup//user.slice/user-1000.slice/session-c1.scope/memory.limit_in_bytes: N/A
> I20241007 14:32:51.190994 825726 dfly_main.cc:438] container limits: read /sys/fs/cgroup//user.slice/user-1000.slice/session-c1.scope/memory.max: max
> I20241007 14:32:51.191002 825726 dfly_main.cc:438] container limits: read /sys/fs/cgroup//user.slice/user-1000.slice/session-c1.scope/memory.high: max
> ```


Yep this makes sense, as I said the paths `/sys/fs/cgroup/memory/memory.max` are probably invalid. With the changes in this PR the output will be:

```
I20241007 14:45:48.500520 133436 dfly_main.cc:459] mem_path = /sys/fs/cgroup//user.slice/user-1000.slice/user@1000.service/app.slice/app-org.gnome.Terminal.slice/vte-spawn-423f3ba1-6da8-42ce-b9bd-77849ef0a079.scope
I20241007 14:45:48.500545 133436 dfly_main.cc:460] cpu_path = /sys/fs/cgroup//user.slice/user-1000.slice/user@1000.service/app.slice/app-org.gnome.Terminal.slice/vte-spawn-423f3ba1-6da8-42ce-b9bd-77849ef0a079.scope
I20241007 14:45:48.500578 133436 dfly_main.cc:438] container limits: read /sys/fs/cgroup/memory/memory.limit_in_bytes: N/A
I20241007 14:45:48.500612 133436 dfly_main.cc:438] container limits: read /sys/fs/cgroup/memory/memory.max: N/A
I20241007 14:45:48.500638 133436 dfly_main.cc:438] container limits: read /sys/fs/cgroup/memory.limit_in_bytes: N/A
I20241007 14:45:48.500659 133436 dfly_main.cc:438] container limits: read /sys/fs/cgroup/memory.max: N/A
I20241007 14:45:48.500681 133436 dfly_main.cc:438] container limits: read /sys/fs/cgroup/user.slice/memory.limit_in_bytes: N/A
I20241007 14:45:48.500723 133436 dfly_main.cc:438] container limits: read /sys/fs/cgroup/user.slice/memory.max: max
I20241007 14:45:48.500752 133436 dfly_main.cc:438] container limits: read /sys/fs/cgroup//user.slice/user-1000.slice/user@1000.service/app.slice/app-org.gnome.Terminal.slice/vte-spawn-423f3ba1-6da8-42ce-b9bd-77849ef0a079.scope/memory.limit_in_bytes: N/A
I20241007 14:45:48.500784 133436 dfly_main.cc:438] container limits: read /sys/fs/cgroup//user.slice/user-1000.slice/user@1000.service/app.slice/app-org.gnome.Terminal.slice/vte-spawn-423f3ba1-6da8-42ce-b9bd-77849ef0a079.scope/memory.max: max
I20241007 14:45:48.500828 133436 dfly_main.cc:438] container limits: read /sys/fs/cgroup//user.slice/user-1000.slice/user@1000.service/app.slice/app-org.gnome.Terminal.slice/vte-spawn-423f3ba1-6da8-42ce-b9bd-77849ef0a079.scope/memory.high: max

```

Also keep in mind that the problem described in the issue was not `debug logs` you see here but rather triggering:

```
  555     LOG(ERROR) << "Failed in deducing any cgroup limits with paths " << mem_path << " and "
  556                << cpu_path;

```

Which should be now fixed since `/sys/fs/cgroup/user.slice` should contain the controllers required

**Review by adiholden — APPROVED** [review]

**Pull request merged**